File: edu/two_sum.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;
class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        vector<int> result;
        unordered_map<int, int> m;
        for (int i = 0; i < nums.size(); i++)
        {
            int key = target - nums[i];
            if (m.find(key) != m.end())
            {
                return {m[key], i};
            }

            if (m.find(nums[i]) == m.end())
            {
                m[nums[i]] = i;
            }
        }
        return result;
    }
};

int main()
{
    vector<int> v = {4, 5, 6};
    int target = 10;
    Solution s;
    vector<int> result = s.twoSum(v, target);
    cout << result[0] << " " << result[1] << endl;
}

File: Edu/Two Sum.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        const map = new Map();

        for (let i = 0; i < nums.length; i++) {
            const diff = target - nums[i];
            if (map.has(diff)) {
                return [map.get(diff), i];
            }
            if (!map.has(nums[i]))
                map.set(nums[i], i);
        }
        return [];
    }

}

const sol = new Solution();
const nums = [5, 5, 3], target = 8;
console.log(sol.twoSum(nums, target));
